[RAD-2440] Unit tests for full GRPC server

mark-idleman opened the pull request:
Removes original server unit tests that were spinning up the old HTTP servers + hitting them with requests, and replaces them with similar tests that operate on the GRPC servers we actually use in production. Next steps will be testing the actual underlying server implementation (instead of using the servers as a black box and testing end-to-end querying + results), ~and of course getting Circle to run these tests [automatically](https://replicahq.atlassian.net/browse/RAD-2548)~.

-  Adds mini_kc test data for use in unit tests. It turns out that mini_kc is _so_ small that very few ODs actually work for transit requests; I've found a specific OD that does return a full-fledge transit route, so I stuck with that for these tests
- Tests added so far are simple, but check quite a lot. I mostly focused on ensuring that all fields in response were included + made sense heuristically (where applicable), with a few tests also checking behavior when ODs are "bad" (outside of the region's bbox). It would also be nice to test "bad" points within the bbox - ones that Graphhopper can snap to the road network, but can't find routes for, but given how small the mini_kc cutout is, I'm not sure they exist here
- GRPC test server was spun up following guidance from [here](https://stackoverflow.com/questions/37552468/how-to-unit-test-grpc-java-server-implementation-functions) and [here](https://grpc.github.io/grpc-java/javadoc/io/grpc/inprocess/InProcessServerBuilder.html), if anyone's curious about the internals
- Automates the running of all tests in CircleCI. Splits up the main dockerfile we use into 2 separate files; one contains all the stages needed to build everything, and gives us access to maven so we can run tests with it; the latter "server" dockerfile runs the final stage that grabs the minimal jars + configs needed to spin up the server, and ends with the `CMD` that runs the server itself. In between these two stages, we run `mvn test` on the first image, as a new step in our CircleCI build

mark-idleman committed:
first pass: switch main unit test to use real GRPC server + mini nor cal test data

mark-idleman committed:
add mini_kc test data; write first transit unit test

mark-idleman committed:
add basic street based query unit tests

mark-idleman committed:
cleanup; add simple tests for bad points

mark-idleman committed:
try automating mvn test in circle

mark-idleman committed:
try circle version 2.1

mark-idleman committed:
try setting env vars before mvn test so build works

mark-idleman committed:
try it another way

mark-idleman committed:
what if we just stopped blocking tests during our usual mvn build?

mark-idleman committed:
try this again, but pass mvn settings file to tes

mark-idleman committed:
add debug mode

mark-idleman committed:
set mvn settings file + src dir manually

mark-idleman committed:
bump orb version

mark-idleman committed:
get rid of src dir declaration

mark-idleman committed:
just specify settings file, b/c env vars should already be set within circle

mark-idleman commented:
OK, I'm actually trying to get circle to do `mvn test` for us in this PR. One of these attempts is bound to work....(right?)

mark-idleman committed:
try again

mark-idleman committed:
try explicit test command

mark-idleman committed:
turn off verify dependencies

mark-idleman committed:
try another way, not using orb (separate job that runs mvn command directly)

mark-idleman committed:
use correct jdk to build

mark-idleman committed:
try another image

mark-idleman committed:
and again

mark-idleman committed:
try #12345

marcopalmeri commented on web/src/test/java/com/replica/RouterServerTest.java:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    @BeforeAll
+    public static void setUp() throws Exception {
```
@mark-idleman - It might be worth factoring this out into some test utils or something, so it can be reused in other tests?

mark-idleman replied:
Yeah, I think that's definitely a good idea, I was just debating doing it now vs. in the next change because we still don't have very many tests. But, might be nice to just combine now, I'll see how much I can move to a shared test util(s)

mark-idleman replied:
Refactored and pulled out all shared logic into a new class that I had the individual tests extend from. This also meant that I moved the existing tests @danielhfrank added recently for the street export to using mini_kc's OSM + GTFS, instead of the original testing files; I'm not removing those files, though, because they could be useful in the future if we want to test weird GTFS edge cases or something

mark-idleman committed:
try another way

mark-idleman committed:
minor swap

mark-idleman committed:
Revert "minor swap"

This reverts commit ebd793035ed3ef14a31abfea976ee4f9d76b360a.

mark-idleman committed:
don't git ignore test data

mark-idleman committed:
oops, forgot to add the mini kc osm

mark-idleman committed:
fix copy to point to parent image

mark-idleman committed:
fix COPY again

mark-idleman commented:
There are about 1000 ways to get tests _almost_ working in Circle given all the Docker magic we're already doing, but I finally found a way that _actually_ works. Our Circle workflow now has a step where it runs `mvn test` after all of the code is built, which required splitting our current dockerfile into 2 so we can run tests when we still have access to maven in the image we're constructing. There are probably ways to make tests a separate github check (ie a separate Circle job, instead of a step of our current build job), but it's messy. I'll leave it for another day.

mark-idleman committed:
remove tar of gtfs

mark-idleman committed:
Revert "remove tar of gtfs"

This reverts commit 510839b308b07fc08134dc03b66b4c78372f52dc.

mark-idleman committed:
remove untarred gtfs instead

danielhfrank commented:
I think this looks promising! I don't have a lot to say about the application logic of the tests you added, it all seems reasonable to me. A couple of things about the image-publishing changes though:
* can you push up a revision with a change that makes the tests fail, to confirm that the circleci job fails?
* besides the circleci job succeeding, how can we confirm that the router will still run as expected after the docker changes?

mark-idleman committed:
Merge branch 'original-direction' into grpc_unit_tests

mark-idleman committed:
refactor to use shared test class

mark-idleman committed:
close graphhopper after each test

mark-idleman committed:
fix tests

mark-idleman committed:
update class path for RouterServer in dockerfile RUN

mark-idleman commented:
@danielhfrank good questions! 

> can you push up a revision with a change that makes the tests fail, to confirm that the circleci job fails?

Definitely! In the last few commits I've pushed I've actually inadvertently done this (eg [here](https://app.circleci.com/pipelines/github/replicahq/graphhopper/622/workflows/a7e2da94-601e-4106-b807-a15a3f11a6dc/jobs/658)), and it's correctly caused the circle job to fail

> besides the circleci job succeeding, how can we confirm that the router will still run as expected after the docker changes?

The easiest way is a bit manual - building a server with the new base image and trying it out. I've [done that](https://cloud.prefect.io/replica/flow-run/b55fd495-7283-4fed-b9f3-bff2aa4c5b1c) with mini nor cal, building a new router that can be spun up with `docker run -it -p 8998:8998 -p 50051:50051 us.gcr.io/model-159019/mininorcal-gh-router-service:bd740266fa8a66db5ecaa61bf8b18f42994998e2`. I can successfully request routes within the GUI, which _should_ be enough to know that both the backend + frontend components were installed correctly during the Docker build (I actually caught a bug this way, that I corrected with my latest commit)
<img width="1878" alt="Screen Shot 2021-06-29 at 10 35 31 AM" src="https://user-images.githubusercontent.com/13446427/123842535-be6e7900-d8c5-11eb-993b-a30c70b349e2.png">
A slightly better approach (in a future ticket) will be adding a true integration test, that builds a small router using the base docker image constructed at the end of the Circle CI job, and hits it with requests (similar to the tests I added in this PR, but not hitting an internally-instantiated service)

kaelgreco reviewed (approved):
This is so great and it makes me feel 100x better about the repo.

It might make sense to transition to `mini_nor_cal` over `mini_kc` for test data since `mini_kc` is so small and has some idiosyncrasies due to its manual construction. Not necessary now.

It'd also be nice to add basic info to the repo readme to explain testing framework (how to run all /class/single). but it probably makes more sense to do when we include other java build/run info that's currently in the python repo.

it'd be cool to get some semblance of test coverage in place and report +/- on commit too

mark-idleman commented on test_gh_config.yaml:
```diff
@@ -0,0 +1,52 @@
+# Configuration file for standard GraphHopper routers.
```
The main difference is the GTFS + OSM are hardcoded in the test config, whereas they're left blank in the default config. With the default config, we can pass the GTFS + OSM paths at runtime using Dropwizard arguments, but I wasn't able to get something similar working with the programatic CLI commands that we run during test setup here, so I did the next best thing and just created a config where they're hardcoded

mark-idleman commented on web/src/test/java/com/replica/RouterServerTest.java:
```diff
@@ -0,0 +1,267 @@
+        // Check that PT leg contains proper info
+        RouterOuterClass.PtLeg ptLeg = path.getPtLegs(0);
+        assertTrue(ptLeg.getArrivalTime().getSeconds() > ptLeg.getDepartureTime().getSeconds());
+        assertTrue(ptLeg.getStableEdgeIdsCount() > 0); // check that the GTFS link mapper worked
```
🎉

JacobHayes reviewed (commented):
Dockerfile changes LGTM. Will we need to update the `model` side to use the new `com.replica.` prefix in the java commands?

mark-idleman committed:
switch to correct validity start date; add comment with validity date + bbox

mark-idleman commented on web/src/test/java/com/replica/RouterServerTest.java:
```diff
@@ -0,0 +1,267 @@
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class RouterServerTest extends ReplicaGraphHopperTest {
+    private static final Timestamp EARLIEST_DEPARTURE_TIME =
+            Timestamp.newBuilder().setSeconds(Instant.parse("2017-07-21T08:25:00Z").toEpochMilli() / 1000).build();
```
Hmmm good call, I'll add a small note with the official mini_kc validity start date + a hint at the bbox. Actually, we weren't using the most up-to-date date here anyway, so I updated it to match with the current validity start date in mini_kc

mark-idleman commented:
@kaelgreco 100% agree that `mini_nor_cal` would be better here, and that's what I started with, but the OSM is too big for github to let you push it :( We'd have to set something up so it would be downloaded from GCS at test startup or something, annoying but do-able.

re: your other suggestions on adding info to the readme (eventually in this repo, not model) and getting a more formal report on test coverage/etc, I also completely agree. There's a lot more to add to this basic foundation, but just having tests that run every commit feels very good to start!

@JacobHayes I don't think so! The only thing that was moved to a new package was the server itself, and the only spot that needs updating with the `com.replica` prefix for the server is in the Dockerfile itself, which I've updated in this PR. The other commands haven't seen any changes here, and I guess we got confirmation that they still work when I was able to build mini nor cal completely with a base image from this branch 😁
